### ISSUE TITLE
IMap.put(String,Object,long,TimeUnit) broken starting with 3.9.2

### ISSUE THREAD

**Issue by carstenartur** [hazelcast/hazelcast#13272]
After ttl exceeded the  map entry is still there with all hazelcast versions since 3.9.2. 
3.9.1 and earlier are ok.

**Comment by ahmetmircik** [comment]
@carstenartur can you please share a reproducer? i tried but didn't manage to reproduce the failure.

**Comment by carstenartur** [comment]
added a test as https://github.com/hazelcast/hazelcast/pull/13278

**Comment by jerrinot** [comment]
@carstenartur: thanks for the reproducer, I'll have a look shortly!

**Comment by jerrinot** [comment]
@carstenartur: ok, that's a good reproducer, thanks! 

so apparently Hazelcast not expire locked entries. When I remove the locking from the test case then it's all OK.

the question is whether this is desirable. @ahmetmircik @vbekiaris ?

**Comment by carstenartur** [comment]
Thanks, and keep in mind that it expires using 3.9.1 and earlier. The change in the semantic is not documented.

**Comment by jerrinot** [comment]
I just tried your test-case against 3.9.1 and 3.9 and it fails with these versions too. I see no behaviour change you are reporting. In the fact I see the logic for not expiring locked keys has been in the codebase for quite a long time. 

This is from the [3.5.x serie](https://github.com/jerrinot/hazelcast/blob/v3.5.2/hazelcast/src/main/java/com/hazelcast/map/impl/AbstractEvictableRecordStore.java#L302)

Can you please validate your setup? Thanks!

**Comment by jerrinot** [comment]
hello @carstenartur,

I am still confused. I tried your newer version of the test:
```java
@Test
    public void afterLongerExpirationEntryShouldBeAway() throws InterruptedException {
        final String KEY = "key";

        Config config = new Config()
                .setInstanceName("instance")
                .addMapConfig(new MapConfig().setName("test").setMaxSizeConfig(new MaxSizeConfig(200,MaxSizeConfig.MaxSizePolicy.FREE_HEAP_SIZE)).setTimeToLiveSeconds(20));
        final HazelcastInstance node = createHazelcastInstance(config);
        try {
            IMap map = node.getMap("test");
            /**
             * after 1 second entry should be evicted
             */
            map.put(KEY, "value", 1, TimeUnit.SECONDS);
            /**
             * short time after adding it to the map, all ok
             */
            map.lock( KEY );
            Object object = map.get( KEY );
            map.unlock( KEY );
            assertNotNull( object );

            Thread.sleep( 3600 );
            /**
             * More than one second after adding it, now it should be away
             */
            map.lock( KEY );
            object = map.get( KEY );
            map.unlock( KEY );
            assertNull( object );
        }finally {
            node.shutdown();
        }
    }
```

and it's still failing on 3.9.1 and after exploring this a bit more I am now convinced this is an intended behaviour: We do not want to remove entries while somehow is holding a lock on them. 

Also I am not quite sure what SW you are referring to by version `2.7.5`. We have never released this version.

**Comment by carstenartur** [comment]
You find this version after you fork the hazelcast git repository to you local eclipse:
![grafik](https://user-images.githubusercontent.com/3164220/41221481-88279a72-6d64-11e8-8755-98800a0a4ffa.png)
There are no newer tags available for whatever reason. 
Here is the commit that adds tests to 2.7.5 where then the test with the longer waittime does *not* fail while it fails on current hazelcast:
https://github.com/carstenartur/hazelcast/commit/5b3c4460f03986a8124188c0f149b8aa0ee5158d
I am not sure what this all means. Is locking really meant to hold on eviction? Is the test code wrong because it causes a race condition? Is the documentation really reflecting reality?

**Comment by jerrinot** [comment]
I am sorry, I still do not understand what you mean by version 2.7.5. I don't even see it on the screenshot. Do you mean 3.7.5? 

In Git you have to explicitly use the `--tags` to fetch tags from a remote repository.

**Comment by carstenartur** [comment]
ah, yes, I mean 3.7.5 :)

**Comment by carstenartur** [comment]
Just tried as advised to fetch the tags from remote repository and checked with the 3.9.1 checkout. As reported the test using the longer waittime does *not* fail using this version. So after 3.6 seconds the map entry is evicted (as it should imho). As the lock only is active for the time pulling the map entry I think the lock should not have an impact anyway. I think the eviction is the right thing in this situation and the question is why it is not evicted earlier. And of course why behavior changed and eviction does not any longer happen at all in newer hazelcast versions. Seems to be a bug imho.

**Comment by jerrinot** [comment]
ok, I can finally see the behaviour change. Let me have a look what has changed.

**Comment by jerrinot** [comment]
ok, I can see what is going on: There is a periodic task to remove expired keys. 

In 3.9.1 and older the task is scheduled as soon as the `ExpirationManager` instance is created. In [3.9.2+](https://github.com/hazelcast/hazelcast/pull/12019) the task is scheduled only when a member state is set as STARTED. 

There is some delay between the time when the `ExpirationManager` instance is created and member state is STARTED. Hence in 3.9.1 the expiration task kicks in while the test is sleeping and it removed the entry. In 3.9.2 the task kicks in later - and at this time the entry is already locked so it's not removed. 

A long story short: There is a race in both 3.9.2 and older versions. However in 3.9.2 we change the timing a bit. 

I think we should do this: Change the lock operation to check whether an entry is already expired. If it is then it should eagerly evict it before acquiring a lock. With this it won't rely on (racy) background task. It's the same algorithm the Get operation uses. @ahmetmircik, @vbekiaris: WDYT?

@carstenartur: thanks again for a reproducer!

**Comment by carstenartur** [comment]
Thanks for your help. regarding the IMap.put() I suggest to add a remark to the javadoc about accuracy of the timeout value. I was not aware that this is not using a scheduled event for eviction that has the potential to be very precise. A periodic task with fixed time slots to remove expired keys limits the accuracy of the eviction time, right?

**Comment by jerrinot** [comment]
@carstenartur: it's actually more complicated:
When you call `map.get()` then Hazelcast always checks whether an item is expired. If an item is actually expired then Hazelcast will remove it from a record store and the `map.get()` returns `null.` If you think about it then this would have worked even without any periodic task at all. The periodic task is just an optimization to reclaim memory faster - even when there is no `map.get()` call. 

There is a caveat: We never expire an item of a locked key. Once you lock an entry then it shouldn't change before you unlock. 

You can see the behaviour yourself: When you remove the last `lock()/unlock()` pair from your test then it's passing - because the entry is eagerly expired during `map.get()`

So here is my proposal: `map.lock()` operation should behave similarly to `map.get()` operation: It should eagerly remove expired item just before acquiring a lock.

**Comment by vbekiaris** [comment]
> I think we should do this: Change the lock operation to check whether an entry is already expired.

This makes sense to me, so if the key is already expired before the lock is acquired, `map.get(k)` will return `null` while the lock is being held.

**Comment by ahmetmircik** [comment]
@jerrinot it makes sense, only worry is where to do this check. `LockOperation` is a shared operation between services not a map-specific one, may be we can try to do this check inside `DefaultRecordStore`

**Comment by carstenartur** [comment]
Any chances that this will be fixed with other changes in this area currently being worked on?

**Comment by johanhammar** [comment]
Hi!

I think we're affected by this bug. We are using Spring Boot and when upgrading to 2.0.x and then later 2.1.0 we started to see that atleast one of the caches stopped evicting entries. 

Problem is that we can't go back to 3.9.1 or earlier due to Spring Boot, we must back to Spring Boot 1.5.x

Funny thing is that I implemented a MapListener and while "observing" the cache, it seems like cache eviction started working. Timing issue? 

So, is there a workaround for this problem?
